thinc/util.py

```diff
@@ -203,12 +203,22 @@ def copy_array(dst: ArrayXd, src: ArrayXd) -> None:  # pragma: no cover
         numpy.copyto(dst, src)  # type: ignore
 
 
-def to_categorical(Y: IntsXd, n_classes: Optional[int] = None) -> FloatsXd:
+def to_categorical(
+    Y: IntsXd,
+    n_classes: Optional[int] = None,
+    label_smoothing: float = 0.0,
+) -> FloatsXd:
+    assert label_smoothing >= 0 and label_smoothing < 1
     xp = get_array_module(Y)
     if n_classes is None:
         n_classes = int(numpy.max(Y) + 1)  # type: ignore
-    # Unfortunately, cupy does not support put_along_axis.
-    return xp.eye(n_classes, dtype="float32")[Y]
+    label_distr = xp.full((n_classes, n_classes), label_smoothing / (n_classes - 1))
+    xp.fill_diagonal(label_distr, 1 - label_smoothing)
+    if Y.dtype.itemsize == 4:  # type: ignore
```

Comment:
Why do you want to size the floats to the integer width? This changes the API of `to_categorical` (since it would always return `float32` before). But the width of the integer is also not really related to the width of the labels?

If we want to provide flexibility, `to_categorical` could take an optional `dtype` argument and pass that to `xp.full`.

Reply:
I'm not sure why is it supposed to always return `float32`. A bunch of the `ops` in `backend` are using the `like` mechanism right? So if the func gets `float32` it supposed to return `float32` also same for `float64`. Since `to_categorical: Ints1d -> Floats2d` we can't get keep the type, but I was just thinking that the most intuitive behavior is to keep the width. But I'll revert it to the previous version of always returning `float32`.

Reply:
I don't see the integer width as related to the floating point width. Even if you are doing math in `float64` for higher precision, 32-bit integers are fine for most occasions such as class labels, indices, lengths, etc. 32-bit is also the default integer size on e.g. CUDA GPUs.

I agree that it would be nice if the function could also return `float64`, but I think it makes more sense then to add a `dtype` argument. Then a caller could do the right thing based on the information that the caller has.

Reply:
I think its good enough to stay with `float32`, because the only other value we could accept is `float64` atm. (right? :D )

Reply:
Yeah, no surprises. We can always choose to add a `dtype` argument later if necessary.